pkg/limits/frontend/ring.go

```diff
@@ -53,70 +65,129 @@ func (g *RingStreamUsageGatherer) GetStreamUsage(ctx context.Context, r GetStrea
-func (g *RingStreamUsageGatherer) forAllBackends(ctx context.Context, r GetStreamUsageRequest) ([]GetStreamUsageResponse, error) {
+	// Get all healthy instances across all zones.
 	rs, err := g.ring.GetAllHealthy(LimitsRead)
 	if err != nil {
 		return nil, err
 	}
-	return g.forGivenReplicaSet(ctx, rs, r)
-}
-
-func (g *RingStreamUsageGatherer) forGivenReplicaSet(ctx context.Context, rs ring.ReplicationSet, r GetStreamUsageRequest) ([]GetStreamUsageResponse, error) {
-	partitionConsumers, err := g.getPartitionConsumers(ctx, rs.Instances)
+	// Get the partition consumers for each zone.
+	zonesPartitions, err := g.getZoneAwarePartitionConsumers(ctx, rs.Instances)
 	if err != nil {
 		return nil, err
 	}
+	// In practice, we want zones to be queried in random order to spread
+	// reads. However, in tests we want a deterministic order so test cases
+	// are stable and reproducible. When compared to just iterating over
+	// a map, this allows us to achieve both.
+	zonesToQuery := make([]string, len(zonesPartitions))
+	for zone := range zonesPartitions {
+		zonesToQuery = append(zonesToQuery, zone)
+	}
+	slices.SortFunc(zonesToQuery, g.zoneCmp)
+	// Make a copy of the stream hashes as we will prune this slice each time
+	// we receive the responses from a zone.
+	streamHashesToQuery := make([]uint64, len(r.StreamHashes))
+	copy(streamHashesToQuery, r.StreamHashes)
+	// Query each zone as ordered in zonesToQuery. If a zone answers all
+	// stream hashes, the request is satisifed and there is no need to query
+	// subsequent zones. If a zone answers just a subset of stream hashes
+	// (i.e. the instance that is consuming a partition is unavailable or the
+	// partition that owns one or more stream hashes does not have a consumer)
+	// then query the next zone for the remaining stream hashes. We repeat
+	// this process until all stream hashes have been queried or we have
+	// exhausted all zones.
+	responses := make([]GetStreamUsageResponse, 0)
+	for _, zone := range zonesToQuery {
+		result, streamHashesToDelete, err := g.getStreamUsage(ctx, r.Tenant, streamHashesToQuery, zonesPartitions[zone])
+		if err != nil {
+			continue
+		}
+		responses = append(responses, result...)
+		// Delete the queried stream hashes from the slice of stream hashes
+		// to query. The slice of queried stream hashes must be sorted so we
+		// can use sort.Search to subtract the two slices.
+		slices.Sort(streamHashesToDelete)
+		streamHashesToQuery = slices.DeleteFunc(streamHashesToQuery, func(streamHashToQuery uint64) bool {
+			// see https://pkg.go.dev/sort#Search
+			i := sort.Search(len(streamHashesToDelete), func(i int) bool {
+				return streamHashesToDelete[i] >= streamHashToQuery
+			})
+			return i < len(streamHashesToDelete) && streamHashesToDelete[i] == streamHashToQuery
+		})
+		// All stream hashes have been queried.
+		if len(streamHashesToQuery) == 0 {
+			return responses, nil
+		}
+	}
+	// Treat remaining stream hashes as unknown streams.
```

Comment:
I think there is an interesting question here of is this the correct behavior, or should this be an error?

Reply:
In the case where some subset of stream hashes cannot be queried because either:

1. There are no pods assigned as consumers for the partition that owns the stream hashes
2. All pods across all zones that consume this partition were unavailable

This subset of stream hashes will be treated as an unknown stream. Is this the correct thing to do?

Reply:
I believe the correct behavior here is to return an error because:
1. We classify only two non ambiguous types or unknown streams: `exceeds_max_streams` and `exceeds_rate_limit`. Adding a third one for "could-not-query` is too ambiguous.
2. Even if we return the streams as unknown back it is not easy to tell the distributor clients what to do. Should they just retry? Or bubble up an error to the agents?